pygments/lexers/ul4.py

```diff
@@ -0,0 +1,261 @@
+            ),
+            (
+                # Template header with name (potentially followed by the signature):
+                # ``<?ul4 foo(bar=42)?>``
+                r"(<\?)(\s*)(ul4)(\s*)([a-zA-Z_][a-zA-Z_0-9]*)?\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword, token.Text, token.Token.Name.Function),
+                "ul4", # Switch to "expression" mode
+            ),
+            (
+                # Comment:
+                # ``<?note foobar?>``
+                r"<\?\s*note\s+.*?\?>",
+                token.Comment,
+            ),
+            (
+                # Template documentation:
+                # ``<?doc foobar?>``
+                r"<\?\s*doc\s+.*?\?>",
+                token.String.Doc,
+            ),
+            (
+                # ``<?ignore?>`` tag for commenting out code:
+                # ``<?ignore?>...<?end ignore?>``
+                r"<\?\s*ignore\s*\?>",
+                token.Comment,
+                "ignore", # Switch to "ignore" mode
+            ),
+            (
+                # ``<?def?>`` tag for defining local templates
+                # ``<?def foo(bar=42)?>...<?end def?>``
+                r"(<\?\s*)(def)(\s*)([a-zA-Z_][a-zA-Z_0-9]*)?\b",
+                lexer.bygroups(token.Comment.Preproc, token.Keyword, token.Text, token.Token.Name.Function),
+                "ul4", # Switch to "expression" mode
+            ),
+            (
+                # The rest of the supported tags
+                r"(<\?)(\s*)(printx|print|for|if|elif|else|while|code|renderblocks?|render)\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword),
+                "ul4", # Switch to "expression" mode
+            ),
+            (
+                # ``<?end?>`` tag for ending ``<?def?>``, ``<?for?>``,
+                # ``<?if?>``, ``<?while?>``, ``<?renderblock?>`` and
+                # ``<?renderblocks?>`` blocks.
+                r"(<\?)(\s*)(end)\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword),
+                "end", # Switch to "end tag" mode
+            ),
+            (
+                # ``<?whitespace?>`` tag for configuring whitespace handlng
+                r"(<\?)(\s*)(whitespace)\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword),
+                "whitespace", # Switch to "whitespace" mode
+            ),
+            # Plain text
+            (r"[^<]+", token.Token.Other),
+            (r"<", token.Token.Other),
+        ],
+        # Ignore mode ignores everything upto the matching ``<?end ignore?>`` tag
+        "ignore": [
+            # Nested ``<?ignore?>`` tag
+            (r"<\?\s*ignore\s*\?>", token.Comment, "ignore"),
+            # ``<?end ignore?>`` tag
+            (r"<\?\s*end\s+ignore\s*\?>", token.Comment, "#pop"),
+            # Everything else
+            (r".+", token.Comment),
```

Comment:
Are you sure that this works? `+` matches greedily as much as it can, until a newline here because `.` matches everything but a newline. I think this should be `.+?`. Your test file appears to work because the `<?end ignore?>` tag appears at the beginning of the line.

Reply:
You're right, it doesn't work. Pygments doesn't recognize the `<?end ignore?>` tag.

But changing `r".+"` to `r".+?"` gives me
```
'<?ignore?>'  Comment
'\n'          Comment

'\t'          Comment
'I'           Comment
'g'           Comment
'n'           Comment
'o'           Comment
'r'           Comment
'e'           Comment
' '           Comment
...
```

which, I guess, is not what we want. But if I use
```python
            # Everything else
            (r"[^<]+", Comment),
            (r".", Comment),
```
this seems to work properly:

```
<?ignore?>
	Ignore <?this?>
	<?ignore?>
		... and <?this?>
	<?end ignore?>
	... and <?this?>
... and even <?this?><?end ignore?>
<?code hint = "ignored no longer"?>
```
gives me
```
'<?ignore?>'  Comment
'\n\tIgnore ' Comment
'<'           Comment
'?this?>\n\t' Comment
'<?ignore?>'  Comment
'\n\t\t... and ' Comment
'<'           Comment
'?this?>\n\t' Comment
'<?end ignore?>' Comment
'\n\t... and ' Comment
'<'           Comment
'?this?>\n... and even ' Comment
'<'           Comment
'?this?>'     Comment
'<?end ignore?>' Comment
'\n'          Other

'<?'          Comment.Preproc
'code'        Keyword
' '           Text.Whitespace
'hint'        Name
...
```

So this is fixed in https://github.com/doerwalter/pygments/commit/7661beefada985782266680f1f8df8bef7756837

Reply:
Yeah, you're right, that `.+?` regex was nonsensical.